### DISCUSSION TITLE
When running two different LLMs, Editor crashes when exiting playmode

### DISCUSSION THREAD

**Discussion by WillMiles** [undreamai/LLMUnity#237] [Q&A] [answered]
Hello again, I have a new problem. I'm trying to run two separate LLM assets at the same time, so I can use two separate models that are finetuned for different things. However every time I exit playmode it crashes my entire editor. Am I doing something wrong?
Here's the editor.log after the game crashes: https://pastebin.com/tT05xJ28
Also here's the set up for the two LLMs. They both have an LLM character, that is shown below. Is there something I'm doing wrong here?
![image](https://github.com/user-attachments/assets/06c9d5ec-5dae-46a1-915c-805fab3e671e)
![image](https://github.com/user-attachments/assets/e3e570ac-3abf-46c3-b4f5-06d3924f5e49)
![image](https://github.com/user-attachments/assets/9a7dc4ce-2b32-4b3a-948f-0bd103ccd5f1)
![image](https://github.com/user-attachments/assets/4cbc0425-2005-4985-90d7-374b9248319c)

**Comment by WillMiles** [comment]
Hmm it seems after further investigation that doing it this way may not work. Is there an easy way to have multiple models loaded at the same time or do I have to regenerate the LLM whenever I want to switch?
Edit: Nevermind, you explicitly mention  multiple LLM objects in the Readme so now I'm more confused.

**Comment by amakropoulos** [comment]
It should work, I even have a unit test with 2 LLMs with a LLMCharacter each.
You can test that with the MultipleCharacters sample:
- create a second LLM e.g. LLM2 (you can just copy the existing one)
- assign LLM2 to LLMCharacter2
- select models for LLM and LLM2
- run the sample

**Reply by WillMiles** [reply]
Doing exactly as you said above crashes the editor instantly when entering playmode. (I tested the sample without the second LLM and it works just fine). Here's the log from Editor.Log:
https://pastebin.com/YvtAWy9g

**Reply by WillMiles** [reply]
If I assign both characters to the same LLM, but still have the second LLM in the scene, I can enter playmode, but when I leave playmode it crashes again. (This is in the MultipleCharacters sample by the way):
https://pastebin.com/VvUAKnRg

**Reply by amakropoulos** [reply]
Strange.. Can you try setting the context size to something small e.g. 512 in both LLMs?

**Reply by WillMiles** [reply]
Unfortunately that didn't change anything. Still crashes on starting playmode when characters have different LLMs and on leaving playmode when they have the same one.

**Reply by amakropoulos** [reply]
I see this error in the log which may mean that you get some memory error:
Request Asset Garbage Collect because used heap size increased from 511.4 MB to 1.15 GB, exceeding the threshold of 256.0 MB

Which Unity version do you have?
Could you also try by running 2 Qwen2 0.5B models?

**Reply by WillMiles** [reply]
Unity version 2022.3.29f1
Will try .5B model.

**Reply by WillMiles** [reply]
With .5 B model unfortunately still doesn't work. However now it crashes on hitting playmode regardless of which LLM is assigned to which character.

**Reply by WillMiles** [reply]
Something happened and now it always crashes on starting playmode regardless of LLM assignment or which model is selected: https://pastebin.com/N1RCjKkh
Only in the sample though. In the actual game, it still only crashes when exiting playmode, same as before.

**Reply by amakropoulos** [reply]
What happens if you disable this line?
https://github.com/undreamai/LLMUnity/blob/f7b161de0bf48e43b4011f39f1c4be919e2f54a0/Runtime/LLMLib.cs#L363

**Reply by WillMiles** [reply]
Still crashes unfortunately, same situation. Interestingly, that's not line 363 for me, that's line 344. Let me double check if I'm fully updated.

**Reply by WillMiles** [reply]
Okay I was on 2.2.1 apparently. Just updated to 2.2.2. Crash still occurs in the same way, even when disabling that line (which is now line 363). Also, my coworker tried it on their computer and it still crashed so it's independent of the PC apparently.

**Reply by amakropoulos** [reply]
I confirmed that as well on my Windows PC 👍.
I will look into it!

**Reply by WillMiles** [reply]
Looks like you fixed it maybe according to the pull requests?

**Reply by amakropoulos** [reply]
Yes! https://github.com/undreamai/LLMUnity/pull/242 fixes it.
I'm finishing another feature and will release it tomorrow.

**Reply by WillMiles** [reply]
After the update a coworker of mine is having trouble. It's weird because it worked just fine for me. Is there a way to force "update" the LLM service, because I think that's the issue. When I downloaded it the first time it showed a little set up progress bar but not on his. Unfortunately we can't wait until tomorrow for the update, thank you for any support you can give and for all you've done with this package it truly is amazing.
Here's his log:
![image](https://github.com/user-attachments/assets/dec56d7d-96dd-4783-83f8-a2e2066c60d1)

**Reply by WillMiles** [reply]
Ok we figured out that if you click Use extras it sets up the library again but that didn't fix it, same error.

**Reply by amakropoulos** [reply]
you can remove the undream* folder from Assets/StreamingAssets and restart Unity

**Reply by WillMiles** [reply]
Still doesn't work. Even after that it still says "LLM service couldn't be created" Could it be because he is on Windows 11 or is there something else we can try?

**Reply by WillMiles** [reply]
Even a build I sent him didn't work. We think it might have something to do with Windows 11.

**Reply by amakropoulos** [reply]
could you list the contents of the Assets\StreamingAssets\undream* folder?

**Reply by WillMiles** [reply]
![image](https://github.com/user-attachments/assets/4467c048-84f0-4991-b749-4869264f3691)
This is what the build looks like that did work on my (Windows 10) PC and didn't work on his (Windows 11) PC. I will get the one from his assets folder in a moment, but he's redownloaded it multiple times.

**Reply by amakropoulos** [reply]
what is the version in his undreamai folder i.e. undreamai-**v1.1.10**-llamacpp

**Reply by Eelguezabal** [reply]
Hello, I am the coworker, again thank you again for your help

This is what my folder looks like
<img width="208" alt="Screenshot 2024-09-11 175828" src="https://github.com/user-attachments/assets/7be16e81-8c45-4990-b8c5-8e99efae93b5">

"undreamai-v1.1.12-llamacpp" is the version of the undreamai folder

**Reply by amakropoulos** [reply]
v1.1.12 is still in development :slightly_smiling_face: .

it should work if you use only the changes of https://github.com/undreamai/LLMUnity/pull/242 over the `main` branch.

Alternatively you can:
- set the version to "v1.1.10" here:
https://github.com/undreamai/LLMUnity/blob/a2e19be79480e362740eac45abe3d46232146ca3/Runtime/LLMUnitySetup.cs#L90
- comment out this line:
https://github.com/undreamai/LLMUnity/blob/a2e19be79480e362740eac45abe3d46232146ca3/Runtime/LLMLib.cs#L326
- delete the `Assets/StreamingAssets/undream*` folder and let it download again

**Reply by WillMiles** [reply]
Unfortunately that didn't seem to work either.

**Reply by amakropoulos** [reply]
what errors do you see?

**Reply by WillMiles** [reply]
Ok for some reason it works now. Thank you so much for your help!

**Reply by amakropoulos** [reply, marked as answer]
Thank you for the issue report :pray: ! The fix is now released in v2.2.3.
